jerry-core/api/jerry-snapshot.h

```diff
@@ -24,18 +24,30 @@
 typedef struct
 {
   /* The size of this structure is recommended to be divisible by
-   * JMEM_ALIGNMENT. Otherwise some bytes after the header are wasted. */
+   * uint32_t alignment. Otherwise some bytes after the header are wasted. */
   uint32_t version; /**< version number */
+  uint32_t configuration; /**< configuration flags */
   uint32_t lit_table_offset; /**< offset of the literal table */
-  uint32_t lit_table_size; /**< size of literal table */
-  uint32_t is_run_global; /**< flag, indicating whether the snapshot
-                            *   was saved as 'Global scope'-mode code (true)
-                            *   or as eval-mode code (false) */
+  uint32_t number_of_funcs; /**< number of primary ECMAScript functions */
+  uint32_t func_offsets[1]; /**< function offsets (lowest bit: global(0) or eval(1) context) */
 } jerry_snapshot_header_t;
 
 /**
+ * Evaluate this function on the top of the scope chain.
+ */
+#define JERRY_SNAPSHOT_EVAL_CONTEXT 0x1u
+
+/**
  * Jerry snapshot format version
  */
-#define JERRY_SNAPSHOT_VERSION (7u)
+#define JERRY_SNAPSHOT_VERSION (8u)
+
+/**
+ * Snapshot configuration flags.
+ */
+typedef enum
+{
+  JERRY_SNAPSHOT_FOUR_BYTE_CPOINTER = (1u << 0) /**< compressed pointers are four byte long */
+} jerry_snapshot_configuration_flags_t;
```

Comment:
I listed a couple other pieces of information that impact the compatibility between snapshot and engine here:
https://github.com/jerryscript-project/jerryscript/issues/1981

- If I read the code correctly, `PARSER_MAXIMUM_CODE_SIZE` affects the generated snapshot bytecode (depending on whether `PARSER_MAXIMUM_CODE_SIZE <= 65535` is true or not). Should this be part of this configuration bitset?
- `RegExp`: I noticed that when loading a snapshot that has `RegExp` templates while the engine is configured w/o `RegExp` support, it will crash / hit `JERRY_UNREACHABLE ()` (jerry-snapshot.c:~144). It would be nice to have a bit for whether the snapshot contains any `RegExp` templates and check the bit before attempting to load the snapshot.
- Endianness: at the JrS2017 face to face, you mentioned this is implicitly part of the version field. This is probably fine as it will take a long time before the version bytes become a "palindrome". Maybe it makes sense to point this out in the doc string with the version field.

This may not be an exhaustive list..

Reply:
> when loading a snapshot that has RegExp templates while the engine is configured w/o RegExp support, it will crash

Does this only happen with RegExp? The engine has several other features that can be enabled/disabled independently from each other. Do we need to remember all of them in the snapshot, or it's the RegExp only?

Reply:
> Does this only happen with RegExp?

I don't know for sure... I just happen to run into this problem. There may be more things.

> The engine has several other features that can be enabled/disabled independently from each other. Do we need to remember all of them in the snapshot, or it's the RegExp only?

I think it won't harm, but it doesn't tell you whether the feature is actually *used* / depended upon by the snapshot... In other words, there's a distinction between "snapshot is generated with JrS with regex support enabled" vs "snapshot contains regex template literals". I.e. a snapshot **without** RegEx template literals but generated on JrS **with** RegEx enabled in the build config, can still be executed on a JrS **without** RegEx enabled.

Also, the check for "can I run this snapshot" cannot be a simple "equals" (`==`) check any more then.

Reply:
PARSER_MAXIMUM_CODE_SIZE does not affect the compilation. It is just a user preference, if your environment has a limited space to store the size. It has a maximum though, it is 16M.

Reply:
Regex literal and regex support are different things. It sounds like a good idea to check regex literals.